core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala

```diff
@@ -0,0 +1,140 @@
+  }
+
+  private[tlv] val allFactories: Vector[TLVFactory[TLV]] =
+    Vector(ErrorTLV, PingTLV, PongTLV)
+
+  def fromBytes(bytes: ByteVector): TLV = {
+    val (tpe, _, value) = decodeTLV(bytes)
+
+    allFactories.find(_.tpe == tpe) match {
+      case Some(tlvFactory) => tlvFactory.fromTLVValue(value)
+      case None             => UnknownTLV(tpe, value)
+    }
+  }
+}
+
+sealed trait TLVFactory[+T <: TLV] extends Factory[T] {
+  def tpe: BigSizeUInt
+  def fromTLVValue(value: ByteVector): T
+
+  override def fromBytes(bytes: ByteVector): T = {
+    val (tpe, _, value) = TLV.decodeTLV(bytes)
+
+    require(tpe == this.tpe, s"Invalid type $tpe when expecting ${this.tpe}")
+
+    fromTLVValue(value)
+  }
+}
+
+case class UnknownTLV(tpe: BigSizeUInt, value: ByteVector) extends TLV {
+  require(!TLV.allFactories.map(_.tpe).contains(tpe), s"Type $tpe is known")
+}
+
+object UnknownTLV extends Factory[UnknownTLV] {
+
+  override def fromBytes(bytes: ByteVector): UnknownTLV = {
+    val (tpe, _, value) = TLV.decodeTLV(bytes)
+
+    UnknownTLV(tpe, value)
+  }
+}
+
+/** @see [[https://github.com/lightningnetwork/lightning-rfc/blob/master/01-messaging.md#the-error-message]] */
+case class ErrorTLV(id: ByteVector, data: ByteVector) extends TLV {
+  require(id.length == 32, s"ID associated with error is incorrect length: $id")
+
+  override val tpe: BigSizeUInt = ErrorTLV.tpe
+
+  override val value: ByteVector = {
+    id ++ UInt16(data.length).bytes ++ data
+  }
+}
+
+object ErrorTLV extends TLVFactory[ErrorTLV] {
+  override val tpe: BigSizeUInt = BigSizeUInt(17)
+
+  override def fromTLVValue(value: ByteVector): ErrorTLV = {
+    val id = value.take(32)
+    val len = UInt16(value.drop(32).take(2))
+    val data = value.drop(32 + 2).take(len.toInt)
+
+    ErrorTLV(id, data)
+  }
+}
+
+case class PingTLV(numPongBytes: UInt16, ignored: ByteVector) extends TLV {
+  override val tpe: BigSizeUInt = PingTLV.tpe
+
+  override val value: ByteVector = {
+    numPongBytes.bytes ++ UInt16(ignored.length).bytes ++ ignored
+  }
+}
+
+object PingTLV extends TLVFactory[PingTLV] {
+  override val tpe: BigSizeUInt = BigSizeUInt(18)
+
+  override def fromTLVValue(value: ByteVector): PingTLV = {
+    val numPongBytes = UInt16(value.take(2))
+    val numIgnored = UInt16(value.slice(2, 4))
+    val ignored = value.drop(4).take(numIgnored.toLong)
+
+    PingTLV(numPongBytes, ignored)
+  }
+}
+
+case class PongTLV(ignored: ByteVector) extends TLV {
+  override val tpe: BigSizeUInt = PongTLV.tpe
+
+  override val value: ByteVector = {
+    UInt16(ignored.length).bytes ++ ignored
+  }
+}
+
+object PongTLV extends TLVFactory[PongTLV] {
+  override val tpe: BigSizeUInt = BigSizeUInt(19)
+
+  override def fromTLVValue(value: ByteVector): PongTLV = {
+    val numIgnored = UInt16(value.take(2))
+    val ignored = value.drop(2).take(numIgnored.toLong)
+
+    PongTLV.forIgnored(ignored)
+  }
+
+  def forIgnored(ignored: ByteVector): PongTLV = {
+    new PongTLV(ignored)
+  }
```

Comment:
Are there more types we should define?

Reply:
Do you mean for TLV? We can certainly add more later but these are more or less just to test things out and define conventions.